pkg/drivers/sqlite/sqlite.go

```diff
@@ -81,27 +63,181 @@ func NewVariant(ctx context.Context, driverName, dataSourceName string) (server.
-	//	return nil, nil, errors.Wrap(err, "setup db")
-	//}
 
-	dialect.Migrate(context.Background())
 	if err := dialect.Prepare(); err != nil {
-		return nil, nil, err
+		return nil, nil, errors.Wrap(err, "query preparation failed")
 	}
 
 	return logstructured.New(sqllog.New(dialect)), dialect, nil
 }
 
-func setup(db *sql.DB) error {
-	for _, stmt := range schema {
-		_, err := db.Exec(stmt)
+// setup performs table setup, which may include creation of the Kine table if
+// it doesn't already exist, migrating key_value table contents to the Kine
+// table if the key_value table exists, all in a single database transaction.
+// changes are rolled back if an error occurs.
+func setup(dialect *generic.Generic) error {
+	ctx := context.Background()
+	txn, err := dialect.DB.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer txn.Rollback()
+
+	kineTableCount, _ := countTable(txn, context.Background(), "kine")
+	// Create Kine table if it doesn't already exist
+	if kineTableCount == 0 {
+		if err := createTable(txn); err != nil {
+			return err
+		}
+	}
+
+	if err := migration(txn, ctx, dialect); err != nil {
+		return errors.Wrap(err, "migration failed")
+	}
+	txn.Commit()
+
+	return nil
+}
+
+func createTable(txn *sql.Tx) error {
+	createTableSQL := `CREATE TABLE IF NOT EXISTS kine
+			(
+				id INTEGER PRIMARY KEY AUTOINCREMENT,
+				name TEXT NOT NULL,
+				created INTEGER,
+				deleted INTEGER,
+				create_revision INTEGER NOT NULL,
+				prev_revision INTEGER,
+				lease INTEGER,
+				value BLOB,
+				old_value BLOB
+			)`
+
+	_, err := txn.Exec(createTableSQL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// removeTable drops the table with the given name, in the given transaction.
+func removeTable(txn *sql.Tx, tableName string) error {
+	dropTableSQL := fmt.Sprintf(`DROP TABLE IF EXISTS %s`, tableName)
+	_, err := txn.Exec(dropTableSQL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// alterIndices drops the given old database indices and creates
+// the given new ones.
+func alterIndices(txn *sql.Tx, d *generic.Generic) error {
+	if d.LockWrites {
+		d.Lock()
+		defer d.Unlock()
+	}
+
+	dropIndices := []string{
+		`DROP INDEX IF EXISTS kine_name_index`,
+		`DROP INDEX IF EXISTS kine_name_prev_revision_uindex`,
+	}
+
+	// Drop the old indices
+	for _, stmt := range dropIndices {
+		_, err := txn.Exec(stmt)
 		if err != nil {
 			return err
 		}
 	}
 
+	createIndices := []string{
+		`CREATE INDEX IF NOT EXISTS kine_name_index ON kine (name, id)`,
+		`CREATE UNIQUE INDEX IF NOT EXISTS kine_name_prev_revision_uindex ON kine (prev_revision, name)`,
+	}
+
+	// Create the new indices
+	for _, stmt := range createIndices {
+		_, err := txn.Exec(stmt)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// countTable counts the number of tables with the given name, in the given transaction.
+func countTable(txn *sql.Tx, ctx context.Context, tableName string) (int, error) {
```

Comment:
This function is always used to check if a table exists. It would be better to name this function after that purpose:
 - name it `hasTable`
 - return `tableCount != 0`
 - change the return type to `(bool, error)`
 As a side note, by Go convetion, `ctx` should be the first parameter, not the second.

Reply:
Changed!